File: Classes/PrefsController.m
#import "PrefsController.h"

NSString * const kPreserveBackups = @"PreserveBackups";
NSString * const kAutosave = @"Autosave";
NSString * const kAutosaveInterval = @"AutosaveInterval";
NSString * const kDeleteResourceWarning =  @"DeleteResourceWarning";
NSString * const kLaunchAction = @"LaunchAction";
NSString * const kOpenUntitledFile = @"OpenUntitledFile";
NSString * const kDisplayOpenPanel = @"DisplayOpenPanel";
NSString * const kNoLaunchOption = @"None";

// Transform launch action matrix index to string constants
@implementation PrefsController

+ (NSArray *)launchActions {
    static NSArray *launchActions = nil;
    if (!launchActions)
        launchActions = @[kNoLaunchOption, kOpenUntitledFile, kDisplayOpenPanel];
    return launchActions;
    
}

+ (Class)transformedValueClass {
    return [NSNumber class];
}

+ (BOOL)allowsReverseTransformation {
    return YES;
}

- (id)transformedValue:(id)value
{
    return @([self.class.launchActions indexOfObject:value]);
}

- (id)reverseTransformedValue:(id)value
{
    return self.class.launchActions[[value intValue]];
}

@end

File: Classes/PrefsController.swift
import Foundation

let kConfirmChanges = "ConfirmChanges"
let kDeleteResourceWarning =  "DeleteResourceWarning"
let kLaunchAction = "LaunchAction"
let kOpenUntitledFile = "OpenUntitledFile"
let kDisplayOpenPanel = "DisplayOpenPanel"
let kNoLaunchOption = "None"

// Transform launch action matrix index to string constants
class LaunchActionTransformer: ValueTransformer {
    private static let launchActions: [String] = [
        kNoLaunchOption,
        kOpenUntitledFile,
        kDisplayOpenPanel
    ]
        
    static override func transformedValueClass() -> AnyClass {
        return NSNumber.self
    }
    
    static override func allowsReverseTransformation() -> Bool {
        return true
    }
    
    override func transformedValue(_ value: Any?) -> Any? {
        return Self.launchActions.firstIndex(of: value as! String)
    }
    
    override func reverseTransformedValue(_ value: Any?) -> Any? {
        return Self.launchActions[value as! Int]
    }
}

extension NSValueTransformerName {
    static let launchActionTransformerName = Self(rawValue: "LaunchActionTransformer")
}
